vault/ttl.go

```diff
@@ -0,0 +1,90 @@
+package vault
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/vault/logical"
+)
+
+func calculateTTL(sysView logical.SystemView, increment, backendTTL, period, backendMaxTTL, explicitMaxTTL time.Duration, startTime time.Time) (ttl time.Duration, warnings []string, errors error) {
+	// Truncate all times to the second since that is the lowest precision for
+	// TTLs
+	now := time.Now().Truncate(time.Second)
+	if startTime.IsZero() {
+		startTime = now
+	} else {
+		startTime = startTime.Truncate(time.Second)
+	}
+
+	// Start off with the sys default value, and update according to period/TTL
+	// from resp.Auth
+	ttl = sysView.DefaultLeaseTTL()
+
```

Comment:
With the logic below it's a bit confusing the difference between ttl and increment. I think it really can be calculated up here and is simple:
```go
switch {
case increment > 0:
ttl = increment
case backendTTL > 0:
ttl = backendTTL
default:
ttl = sysView.DefaultLeaseTTL()
}
```